File: karibu-testing-v10-spring/src/main/java/com/github/mvysny/kaributesting/v10/spring/MockSpringServlet.java
package com.github.mvysny.kaributesting.v10.spring;

import com.github.mvysny.kaributesting.v10.mock.MockVaadinHelper;
import com.github.mvysny.kaributesting.v10.Routes;
import com.vaadin.flow.component.UI;
import com.vaadin.flow.function.DeploymentConfiguration;
import com.vaadin.flow.server.ServiceException;
import com.vaadin.flow.server.VaadinServletContext;
import com.vaadin.flow.server.VaadinServletService;
import com.vaadin.flow.spring.SpringServlet;
import kotlin.jvm.functions.Function0;
import org.jetbrains.annotations.NotNull;
import org.springframework.context.ApplicationContext;
import org.springframework.web.context.WebApplicationContext;

import javax.servlet.ServletConfig;
import javax.servlet.ServletException;

/**
 * Makes sure that the {@link #routes} are properly registered,
 * and that {@link MockSpringServletService} is used instead of vanilla {@link com.vaadin.flow.spring.SpringVaadinServletService}.
 * @author mavi
 */
public class MockSpringServlet extends SpringServlet {
    /**
     * The routes registered to the application.
     */
    @NotNull
    public final Routes routes;
    /**
     * The application context.
     */
    @NotNull
    public final ApplicationContext ctx;
    /**
     * Creates new UIs.
     */
    @NotNull
    public final Function0<UI> uiFactory;

    /**
     * Creates new servlet.
     * @param routes The routes registered to the application.
     * @param ctx The application context.
     * @param uiFactory Creates new UIs.
     */
    public MockSpringServlet(@NotNull Routes routes, @NotNull ApplicationContext ctx, @NotNull Function0<UI> uiFactory) {
        super(ctx, false);
        this.ctx = ctx;
        this.routes = routes;
        this.uiFactory = uiFactory;
    }

    @Override
    protected DeploymentConfiguration createDeploymentConfiguration() throws ServletException {
        MockVaadinHelper.mockFlowBuildInfo(this);
        return super.createDeploymentConfiguration();
    }

    @Override
    protected VaadinServletService createServletService(DeploymentConfiguration deploymentConfiguration) throws ServiceException {
        final VaadinServletService service = new MockSpringServletService(this, deploymentConfiguration, ctx, uiFactory);
        service.init();
        routes.register((VaadinServletContext) service.getContext());
        return service;
    }

    @Override
    public void init(ServletConfig servletConfig) throws ServletException {
        super.init(servletConfig);
        // some apps retrieve Spring's WebApplicationContext from the ServletContext,
        // via WebApplicationContext.ROOT_WEB_APPLICATION_CONTEXT_ATTRIBUTE.
        servletConfig.getServletContext().setAttribute(WebApplicationContext.ROOT_WEB_APPLICATION_CONTEXT_ATTRIBUTE, ctx);
    }
}

File: karibu-testing-v10/kt10-tests/src/main/kotlin/com/github/mvysny/kaributesting/v10/spring/MockSpringServlet.kt
package com.github.mvysny.kaributesting.v10.spring

import com.github.mvysny.kaributesting.v10.MockVaadin
import com.github.mvysny.kaributesting.v10.Routes
import com.vaadin.flow.component.UI
import com.vaadin.flow.function.DeploymentConfiguration
import com.vaadin.flow.server.ServiceException
import com.vaadin.flow.server.VaadinRequest
import com.vaadin.flow.server.VaadinServletService
import com.vaadin.flow.server.VaadinSession
import com.vaadin.flow.spring.SpringServlet
import com.vaadin.flow.spring.SpringVaadinServletService
import com.vaadin.flow.spring.SpringVaadinSession
import org.springframework.context.ApplicationContext
import java.util.concurrent.locks.Lock
import java.util.concurrent.locks.ReentrantLock
import javax.servlet.ServletConfig
import javax.servlet.ServletException

/**
 * @author mavi
 */
internal class MockSpringServlet(val routes: Routes, val ctx: ApplicationContext, val uiFactory: () -> UI) : SpringServlet(ctx, false) {
    @Throws(ServletException::class)
    override fun init(servletConfig: ServletConfig) {
        routes.register(servletConfig.servletContext)
        super.init(servletConfig)
    }

    @Throws(ServiceException::class)
    override fun createServletService(deploymentConfiguration: DeploymentConfiguration): VaadinServletService {
        val service: VaadinServletService = object : SpringVaadinServletService(this, deploymentConfiguration, ctx) {
            override fun isAtmosphereAvailable(): Boolean {
                return false
            }

            override fun getMainDivId(session: VaadinSession, request: VaadinRequest): String {
                return "ROOT-1"
            }

            override fun createVaadinSession(request: VaadinRequest): VaadinSession {
                return MockSpringVaadinSession(this, uiFactory)
            }
        }
        service.init()
        return service
    }
}

open class MockSpringVaadinSession(service: SpringVaadinServletService, public val uiFactory: () -> UI) : SpringVaadinSession(service) {
    /**
     * We need to pretend that we have the UI lock during the duration of the test method, otherwise
     * Vaadin would complain that there is no session lock.
     * The easiest way is to simply always provide a locked lock :)
     */
    private val lock: ReentrantLock = ReentrantLock().apply { lock() }

    override fun getLockInstance(): Lock = lock
    override fun close() {
        super.close()
        MockVaadin.afterSessionClose(this, uiFactory)
    }
}
